File: sepQueue.cpp
#include<iostream>
#include<queue>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(): val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x): val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right): val(x), left(left), right(right) {}
};

class Solution {
  public:
  bool recur(TreeNode* left, TreeNode* right) {
    if (!left && !right)
      return true;
    if (!left || !right)
      return false;
    if (left->val != right->val)
      return false;
    return (recur(left->left, right->right) && recur(left->right, right->left));
  }
  bool isSymmetric(TreeNode* root) {
    return recur(root, root);
  }
};

class Solution1 {
public:
  bool isSymmetric(TreeNode* root) {
    if (root == NULL)
      return true;
    if (!root->left && !root->right)
      return true;
    if (!root->left || !root->right)
      return false;

    queue<TreeNode*> queue;
    queue.push(root);
    queue.push(root);
    TreeNode* left;
    TreeNode* right;
    while (!queue.empty())
    {
      left = queue.front();
      queue.pop();
      right = queue.front();
      queue.pop();
      if (left->val != right->val)
        return false;
      if (left->left && right->right)
      {
        queue.push(left->left);
        queue.push(right->right);
      }
      if (left->left || right->right)
        return false;
      if (left->right && right->left)
      {
        queue.push(left->right);
        queue.push(right->left);
      }
      if (left->right || right->left)
        return false;
    }
    return true;
  }
};

int main(int argc, char* argv[]) {
  TreeNode* pval3 = new TreeNode(2);
  TreeNode* pval2 = new TreeNode(2);
  TreeNode* pval1 = new TreeNode(1, pval2, pval3);
  Solution sol;
  Solution1 sol1;
  bool res = sol.isSymmetric(pval1);
  bool res1 = sol1.isSymmetric(pval1);

  cout << res << endl;
  cout << res << endl;

  delete pval3;
  delete pval2;
  delete pval1;
}












// end

File: sep_queue.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSymmetric(self, root):
        def recur(left, right):
            if left==None and right==None:
                return True
            if left==None or right==None:
                return False
            if left.val != right.val:
                return False
            return (recur(left.left, right.right) and recur(left.right, right.left))
        return recur(root, root)

class Solution1:
    def isSymmetric(self, root):
        if root == None:
            return True
        if root.left == None and root.right == None:
            return True
        if root.left == None or root.right == None:
            return False

        queue = []
        queue.append(root)
        queue.append(root)

        left, right = TreeNode(), TreeNode()

        while len(queue):
            left = queue[0]
            queue.pop(0)
            right = queue[0]
            queue.pop(0)

            if left.val != right.val:
                return False
            if left.left and right.right:
                queue.append(left.left)
                queue.append(right.right)
            elif left.left or right.right:
                return False
            if left.right and right.left:
                queue.append(left.right)
                queue.append(right.left)
            elif left.right or right.left:
                return False
        return True

if __name__ == '__main__':
    pval3 = TreeNode(2)
    pval2 = TreeNode(2)
    pval1 = TreeNode(1, pval2, pval3)

    s = Solution()
    s1 = Solution1()

    res = s.isSymmetric(pval1)
    res1 = s1.isSymmetric(pval1)
    print(res)
    print(res1)











# end
